File: ocaml/src/p0061.ml
(* Project Euler: Problem 61 *)

open Core

let make_polygonal_tbl () =
  let make_polygonal_nums i =
    let fn =
      [| (fun _ -> 0) (* not used *)
       ; (fun _ -> 1)
       ; (fun n -> n)
       ; (fun n -> n * (n + 1) / 2) (* triangular number *)
       ; (fun n -> n * n) (* square number *)
       ; (fun n -> n * ((3 * n) - 1) / 2) (* pentagonal number *)
       ; (fun n -> n * ((2 * n) - 1)) (* hexagonal number *)
       ; (fun n -> n * ((5 * n) - 3) / 2) (* heptagonal number *)
       ; (fun n -> n * ((3 * n) - 2)) (* octagonal number *)
      |]
    in
    let aux n =
      let f = fn.(n) in
      Sequence.(
        unfold ~init:1 ~f:(fun i -> Some (f i, succ i))
        |> drop_while ~f:(fun x -> x < 1_000)
        |> take_while ~f:(fun x -> x < 10_000)
        |> to_list)
    in
    aux i
  in
  let tbl = Array.create ~len:9 [] in
  for i = 3 to 8 do
    tbl.(i)
    <- make_polygonal_nums i
       |> List.filter_map ~f:(fun n ->
         let a, b = (n / 100, n mod 100) in
         if b >= 10 then Some (a, b) else None)
  done;
  tbl
;;

let find_chain route_info p_tbl =
  let get_next_chains chain i =
    let key = snd (List.hd_exn chain) in
    List.filter_map
      ~f:(fun ((x, _) as pair) -> if x = key then Some (pair :: chain) else None)
      p_tbl.(i)
  in
  let rec aux chains route =
    match (chains, route) with
    | [], _ -> []
    | _, [] ->
      List.(
        filter chains ~f:(fun lst ->
          let p1, p2 = hd_exn lst in
          let q1, q2 = last_exn lst in
          p1 = q1 && p2 = q2)
        |> map ~f:tl_exn)
    | _, x :: xs -> aux List.(concat (map ~f:(Fun.flip get_next_chains x) chains)) xs
  in
  aux List.(map ~f:singleton p_tbl.(8)) route_info
;;

let compute () =
  let p_tbl = make_polygonal_tbl () in
  let trans_lst =
    List.map ~f:(fun l -> l @ [ 8 ]) (Euler.Util.permutation 5 [ 7; 6; 5; 4; 3 ])
  in

  (*
     From the problem statement:
     - Each elements in cycle is represent by a different number
     - There is only one cycle exist
  *)
  let cycles =
    List.(
      concat_map ~f:(Fun.flip find_chain p_tbl) trans_lst
      |> map ~f:(map ~f:(fun (x, y) -> (100 * x) + y))
      |> filter ~f:(Fun.compose not (contains_dup ~compare:Int.compare)))
  in
  if List.length cycles <> 1
  then failwith "abort"
  else List.(reduce_exn (hd_exn cycles) ~f:( + ))
;;

let solve () = compute () |> Int.to_string

(* Test *)

let%test_unit "Cyclical Figurate Numbers" = [%test_eq: int] (compute ()) 28684

File: rust/src/bin/p0061.rs
// Project Euler: Problem 61

use std::collections::{HashMap, HashSet, VecDeque};

euler::run_solver!(61);

fn solve() -> String {
    compute(8).to_string()
}

fn compute(max_nsides_polygon: usize) -> i64 {
    debug_assert!(max_nsides_polygon > 3);

    let cycles: Vec<Vec<i64>> = find_closed_paths(max_nsides_polygon)
        .into_iter()
        .map(|path| {
            // Change a closed path to a vector of cyclic numbers
            path.windows(2)
                .map(|pair| pair[0] * 100 + pair[1])
                .collect::<Vec<i64>>()
        })
        .filter(|lst| {
            // All numbers in a cycle are different from each other's
            lst.len() == HashSet::<i64>::from_iter(lst.clone()).len()
        })
        .collect();

    // There exists only one cycle
    if cycles.len() == 1 {
        return cycles[0].iter().sum();
    }

    unreachable!();
}

fn find_closed_paths(max_nsides_polygon: usize) -> Vec<Vec<i64>> {
    let mut closed_paths: Vec<Vec<i64>> = Vec::new();
    let polynum_tbl: HashMap<usize, HashMap<i64, Vec<i64>>> = make_polynum_tbl(max_nsides_polygon);
    let stop_condition: u32 = (1 << (max_nsides_polygon + 1)) - 8;

    let mut get_next_states = |(bits, path): (u32, Vec<i64>)| -> Vec<(u32, Vec<i64>)> {
        let mut states: Vec<(u32, Vec<i64>)> = Vec::new();

        // bits: (when max_nsides_polygon = 8)
        //   0b######000
        //     ||||||
        //     |||||+- triangle
        //     ||||+-- square
        //     |||+--- pentagonal
        //     ||+---- hexagonal
        //     |+----- heptagonal
        //     +------ octagonal
        if bits == stop_condition {
            if path[0] == *path.last().unwrap() {
                closed_paths.push(path);
            }
        } else {
            for i in 3_usize..max_nsides_polygon {
                let p_bit = 0b1_u32 << i;
                if bits & p_bit != 0 {
                    continue;
                }
                let next_tbl = polynum_tbl.get(&i).unwrap();
                if let Some(vs) = next_tbl.get(path.last().unwrap()) {
                    for &x in vs {
                        let mut next_path = path.clone();
                        next_path.push(x);
                        states.push((bits | p_bit, next_path));
                    }
                }
            }
        }

        states
    };

    // Search for all closed paths
    let mut q: VecDeque<(u32, Vec<i64>)> = VecDeque::new();
    for (&k, vs) in polynum_tbl.get(&max_nsides_polygon).unwrap() {
        for &v in vs {
            q.push_back((0b1 << max_nsides_polygon, vec![k, v]));
        }
    }
    while !q.is_empty() {
        for (bits, path) in get_next_states(q.pop_front().unwrap()).into_iter() {
            q.push_front((bits, path));
        }
    }

    closed_paths
}

fn make_polynum_tbl(max_nsides_polygon: usize) -> HashMap<usize, HashMap<i64, Vec<i64>>> {
    let mut tbl: HashMap<usize, HashMap<i64, Vec<i64>>> = HashMap::new();
    for p in 3..=max_nsides_polygon {
        let mut x: HashMap<i64, Vec<i64>> = HashMap::new();
        for n in (1_i64..)
            .step_by(p - 2)
            .scan(0, |state, x| {
                *state += x;
                Some(*state)
            })
            .skip_while(|&x| x < 1_000)
            .take_while(|&x| x < 10_000)
            .filter(|&x| x % 100 >= 10)
        {
            x.entry(n / 100).or_default().push(n % 100);
        }
        tbl.insert(p, x);
    }

    tbl
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0061_4() {
        assert_eq!(compute(4), 8181);
    }

    #[test]
    fn p0061_5() {
        assert_eq!(compute(5), 19291);
    }

    #[test]
    fn p0061_8() {
        assert_eq!(compute(8), 28684);
    }
}
